src/ray/common/memory_monitor.cc

```diff
@@ -111,41 +111,120 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetMemoryBytes() {
+  //   Note that `tmpfs` file system might swap data to disk (when option noswap=False)
+  //   but `shmem` value always means the `tmpfs` data size in physical memory.
+  //   We can read `shmem` value from /proc/meminfo `Shmem` item or from
+  //   /sys/fs/cgroup/memory/memory.stat `total_shmem` item.
+  //
+  // Explanation: Why don't use cgroup
+  //  `memory.stat.total_rss` and `memory.stat.inactive_file_bytes` to compute the
+  // in-used memory ?
+  //   In my test using these values can't calculate out correct used memory number,
+  //   cgroup official doc is fuzzy when describing these values.
+  //   But in my test,
+  //   cgroup `memory.usage_in_bytes` value perfectly matches (`MemTotal` - `MemFree`) value,
+  //   so I am sure that `memory.usage_in_bytes` value is correct.
+  //   and cgroup `memory.stat.total_cache` value also perfectly matches
+  //   (`MemAvailable` - `MemFree`) value,
+  //
+  // Correctness testing criteria:
+  //  - [Check cgroup memory.stat value consistency with /proc/meminfo]
+  //    Ensuring the calculated value is consistent with values computed from /proc/meminfo
+  //    Note that cgroup mem file is consistent with /proc/meminfo file unless
+  //    there is bug in cgroup, i.e. we can read MemTotal / MemAvailable / Shmem
+  //    from /proc/meminfo file, then the calculated value by this function
+  //    should equals to `MemTotal` - (`MemAvailable` - `Shmem`)
+  //
+  //  - [OS_managed_cache_and_buffer test]
+  //    Prepare an idle OS environment that has a large number of `free` memory.
+  //    Use dd command to write a large file to disk, after dd  completes,
+  //    the calculated used_memory value should keep nearly the same with the value
+  //    before dd execution.
+  //    But note that free memory will decrease significantly because OS
+  //    cached the file data as part of "OS_managed_cache_and_buffer" I mentioned above.
+  //
+  //  - [/dev/shm test]
+  //    If we use dd command to write a large file to /dev/shm,
+  //    and no swapping occurs (you can use `free -h` to check whether swap size increases),
+  //    after dd completes, the calculated used-memory value should be nearly
+  //    previous_in_used_memory_bytes + bytes_of_written_file
+  //
+  //  - [Host OS SIGKILL signal test]:
+  //    1. get current "used_memory" by running this `GetCGroupV1MemoryUsedBytes` function.
+  //    2. get "swap_space_size" by running `free` command
+  //    3. read "swapped_data_size" value by reading "total_swap" item from /sys/fs/cgroup/memory/memory.stat
+  //    4. Create a program that gradually requests to allocate memory,
+  //       record that after the number "oom_size" memory it get allocated,
+  //       the process is killed by OS SIGKILL signal.
+  //    The "oom_size" recorded in step-(4) should approximately satisfy the following formula:
+  //    oom_size ~== (total_memory + swap_space_size) - used_memory - swapped_data_size
```

Comment:
This part verification is very important.

Note that CGroup official doc is very fuzzy on concrete meanings of values in CGroup mem stat file,

We can do these verifications I listed here to ensure the calculation logic correctness.

Reply:
another related issue worth discussion:

Current Ray memory monitor logic is:

When `used_physical_memory / total_physical_memory > 0.95`, trigger killing Ray tasks.

But this does not fully match Host OS SIGKILL criteria ,
Actually Host OS SIGKILL is triggered by the following criteria:

`(used_physical_memory + used_swap_size) ~== (total_physical_memory + total_swap_size)`

So correspondingly,
Ray memory monitor killing Ray task threshold should be:

`(used_physical_memory + used_swap_size) / (total_physical_memory + total_swap_size) > 0.95`

(We can fix this in follow-up PR)

Reply:
we actually want to kill those processes before the Host OS kills them. It can give us better error message (e.g., host os kills procs using sigkill which we cannot detect).

Reply:
> we actually want to kill those processes before the Host OS kills them

Yes . To accurately implement the goal "kill those processes before the Host OS kills them" , the Memory monitor killing condition should be:

```
(used_physical_memory + used_swap_size) / (total_physical_memory + total_swap_size) > `RAY_memory_usage_threshold`
```

@rkooo567 Is my understanding correct ?

Reply:
Hmm I think it depends on the definition of "~=" in the OS OOM killer. I think if OS oom killer has similar threshold as 95% (including the swap size), the ray memory monitor is useless. Besides, I agree it makes more sense to include swap size in the calculation (though iiuc, most of EC2 disabled swap by default?)